paperqa/agents/models.py

```diff
@@ -0,0 +1,520 @@
+        if answer_tool_name not in v:
+            raise ValueError(
+                f"If using an override, must contain at least the {answer_tool_name}."
+            )
+        return v
+
+
+class ParsingOptions(str, Enum):
+    PAPERQA_DEFAULT = "paperqa_default"
+
+    def available_for_inference(self) -> list[ParsingOptions]:
+        return [self.PAPERQA_DEFAULT]  # type: ignore[list-item]
+
+    def get_parse_type(self, config: ParsingConfiguration) -> str:
+        if self == ParsingOptions.PAPERQA_DEFAULT:
+            return config.parser_version_string
+        assert_never()
+
+
+class ChunkingOptions(str, Enum):
+    SIMPLE_OVERLAP = "simple_overlap"
+
+    @property
+    def valid_parsings(self) -> list[ParsingOptions]:
+        # Note that SIMPLE_OVERLAP must be valid for all by default
+        # TODO: implement for future parsing options
+        valid_parsing_dict: dict[str, list[ParsingOptions]] = {}
+        return valid_parsing_dict.get(self.value, [])
+
+
+class ImpossibleParsingError(Exception):
+    """Error to throw when a parsing is impossible."""
+
+    LOG_METHOD_NAME: ClassVar[str] = "warning"
+
+
+class ParsingConfiguration(BaseModel):
+    """Holds a superset of params and methods needed for each algorithm."""
+
+    ordered_parser_preferences: list[ParsingOptions] = [
+        ParsingOptions.PAPERQA_DEFAULT,
+    ]
+    chunksize: int = 6000
+    overlap: int = 100
+    chunking_algorithm: ChunkingOptions = ChunkingOptions.SIMPLE_OVERLAP
+
+    def chunk_type(self, chunking_selection: ChunkingOptions | None = None) -> str:
+        """Future chunking implementations (i.e. by section) will get an elif clause here."""
+        if chunking_selection is None:
+            chunking_selection = self.chunking_algorithm
+        if chunking_selection == ChunkingOptions.SIMPLE_OVERLAP:
+            return (
+                f"{self.parser_version_string}|{chunking_selection.value}"
+                f"|tokens={self.chunksize}|overlap={self.overlap}"
+            )
+        assert_never()
+
+    @property
+    def parser_version_string(self) -> str:
+        return f"paperqa-{__version__}"
+
+    def is_chunking_valid_for_parsing(self, parsing: str):
+        # must map the parsings because they won't include versions by default
+        return (
+            self.chunking_algorithm == ChunkingOptions.SIMPLE_OVERLAP
+            or parsing
+            in {  # type: ignore[unreachable]
+                p.get_parse_type(self) for p in self.chunking_algorithm.valid_parsings
+            }
+        )
+
+
+class MismatchedModelsError(Exception):
+    """Error to throw when model clients clash ."""
+
+    LOG_METHOD_NAME: ClassVar[str] = "warning"
+
+
+class QueryRequest(BaseModel):
+    query: str = ""
+    id: UUID = Field(
+        default_factory=uuid4,
+        description="Identifier which will be propagated to the Answer object.",
+    )
+    llm: str = "gpt-4o-2024-08-06"
```

Comment:
Can you add a description what this LLM is used for? There's `agent_llm`, `summary_llm`, what do you think of giving this one a more verbose prefix?

Reply:
I agree with this, but since this is used in a lot of places (even outside of paper-qa), I want to make that change in another PR.